File: base/008/solver.cpp
#include <iostream>
#include <map>
#include <vector>
#include <algorithm>
#include <sstream>
#include <iterator>

using namespace std;

class Cliente{
public:
    string name;
    float limite;
    float balance;

    Cliente(string name = "", float limite = 0){
        this->name = name;
        this->limite = limite;
        this->balance = 0;
    }
    friend ostream& operator<<(ostream &os, Cliente cliente){
        os << cliente.name << ":" << cliente.balance << "/" << cliente.limite ;
        return os;
    }
};


class Transaction{
public:
    int id;
    string idCli;
    float value;
    Transaction(int id = 0, string idCli = "", float value = 0.0){
        this->id = id;
        this->idCli = idCli;
        this->value = value;
    }
    friend ostream& operator<<(ostream& os, Transaction tr){
        os << "id:" << tr.id << " " << tr.idCli << ":" << tr.value;
        return os;
    }
};




class Agiota{
    map<string, Cliente> repCli;
    map<int, Transaction> repTr;
    int nextTrId = 0;
    float balance;

    void pushTransacao(string idCli, float value){
        repTr[nextTrId] = Transaction(nextTrId, idCli, value);
        nextTrId++;
    }

public:
    Agiota(float saldo = 0.0){
        this->balance = saldo;
    }

    void addClient(string name, float limite){
        if(repCli.count(name) == 1)
            throw string("fail: cliente ja existe");
        repCli[name] = Cliente(name, limite);
    }

    void lends(string idCli, float value){
        if(value > this->balance)
            throw string("fail: fundos insuficientes");
        if(repCli.count(idCli) == 0)
            throw string("fail: cliente nao existe");
        Cliente& cliente = repCli.at(idCli);
        if(cliente.balance + value > cliente.limite)
            throw string("fail: limite excedido");
        cliente.balance += value;
        this->balance -= value;
        pushTransacao(idCli, value);
    }

    void receive(string idCli, float value){
        Cliente& cliente = repCli.at(idCli);
        if(value > cliente.balance)
            throw string("fail: valor maior que a divida");
        cliente.balance -= value;
        this->balance += value;
        pushTransacao(idCli, -value);
    }



    void kill(string idCli){
        repCli.erase(idCli);
        vector<int> trToRemove;
        for(auto & [k, v] : repTr)
            if(v.idCli == idCli)
                trToRemove.push_back(k);
        for(int k : trToRemove)
            repTr.erase(k);
    }

    friend ostream& operator<<(ostream& os, Agiota& agiota){
        os << "clients:\n";
        for(auto& [k, cli] : agiota.repCli)
            os << "- " << cli << endl;
        os << "transactions:\n";
        for(auto& [k, tr] : agiota.repTr)
            os << "- " << tr << endl;
        os << "balance: " << agiota.balance << endl;
        return os;
    }
};

template<typename T>
T get(stringstream& ss){T t; ss >> t; return t;}

int main(){
    string line, cmd;
    Agiota agiota;
    while(true){
        try{
            getline(cin, line);
            cout << "$" << line << endl;
            stringstream ss(line);
            vector<string> ui(istream_iterator<string>{ss}, istream_iterator<string>());
            string cmd = ui[0];
            if(cmd == "end"){
                break;
            }else if(cmd == "init"){
                agiota = Agiota(stoi(ui[1]));
            }else if(cmd == "addCli"){
                agiota.addClient(ui[1], stoi(ui[2]));
            }else if(cmd == "show"){
                cout << agiota;
            }else if(cmd == "kill"){
                agiota.kill(ui[1]);
            }else if(cmd == "lend"){
                agiota.lends(ui[1], stof(ui[2]));
            }else if(cmd == "receive"){
                agiota.receive(ui[1], stof(ui[2]));
            }else{
                cout << "fail: comando invalido" << endl;
            }
        }catch(string s){
            cout << s << endl;
        }
    }
}

File: base/008/Solver.java
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.TreeMap;
import java.util.stream.Collectors;

class AgException extends RuntimeException {
    public AgException(String message){
        super(message);
    }
}

class Client {
    private String codename;
    private int limite;
    private int balance;

    public Client(String codename, int limite) {
        this.codename = codename;
        this.limite = limite;
        this.balance = 0;
    }
    public String toString() {
        return this.codename + ":" + this.balance + "/" + this.limite;
    }
    String getName() {
        return this.codename;
    }
    void setName(String codename) {
        this.codename = codename;
    }
    int getLimite() {
        return this.limite;
    }
    void setLimite(int limite) {
        this.limite = limite;
    }
    int getBalance() {
        return this.balance;
    }
    void setBalance(int balance) {
        this.balance = balance;
    }
}

class Transaction {
    private int id;
    private String codename;
    private int value;
    public Transaction(int id, String codename, int value) {
        this.id = id;
        this.codename = codename;
        this.value = value;
    }
    public String toString() {
        return "id:" + id + " " + codename + ":" + value;
    }

    int getId() {
        return this.id;
    }
    void setId(int id) {
        this.id = id;
    }
    String getCodename() {
        return this.codename;
    }
    void setCodename(String codename) {
        this.codename = codename;
    }
    int getValue() {
        return this.value;
    }

    void setValue(int value) {
        this.value = value;
    }
}

class Agiota {
    private Map<String, Client> repCli;
    private Map<Integer, Transaction> repTr;
    private int nextTrId;
    private int balance;

    //Guarda a transação e incrementa o nextTrId
    private void pushTransaction(String codename, int value) {
        repTr.put(nextTrId, new Transaction(nextTrId, codename, value));
        nextTrId++;
    }
    //Inicia os atributos
    public Agiota(int balance) {
        this.balance = balance;
        this.repCli = new TreeMap<>();
        this.repTr = new TreeMap<>();
        this.nextTrId = 0;
    }
    //Retorna o cliente ou lança uma exceção se não houver
    public Client getClient(String codename) {
        Client client = repCli.get(codename);
        if(client == null)
            throw new AgException("fail: cliente nao existe");
        return client;
    }
    //Guarda um cliente usando o codenome como chave
    public void addClient(String codename, int limite) {
        Client client = repCli.get(codename);
        if(client != null)
            throw new AgException("fail: cliente ja existe");
        repCli.put(codename, new Client(codename, limite));
    }
    //Empresta dinheiro para o cliente se ainda estiver no limite de crédito do cliente e o agiota possuir
    public void lends(String codename, int value) {
        if(value > this.balance)
            throw new AgException("fail: fundos insuficientes"); 
        Client cliente = this.getClient(codename);
        if(cliente.getBalance() + value > cliente.getLimite())
            throw new AgException("fail: limite excedido");
        cliente.setBalance(cliente.getBalance() + value);
        this.balance -= value;
        pushTransaction(codename, value);
    }
    //Recebe dinheiro do cliente, porém não mais do que a dívida
    public void receive(String codename, int value) {
        Client cliente = getClient(codename);
        if(value > cliente.getBalance())
            throw new AgException("fail: valor maior que a divida");
        cliente.setBalance(cliente.getBalance() - value);
        this.balance += value;
        pushTransaction(codename, -value);
    }
    //Remove o cliente e todas as transações relacionadas a ele
    //Crie uma lista auxiliar com todos os índices a serem removidos
    //e depois remova-os em outro laço. 
    //Sem lista auxiliar, você terá um erro de concorrência
    public void kill(String codename) {
        repCli.remove(codename);
        repTr.entrySet().removeIf(pair -> pair.getValue().getCodename().equals(codename));
    }
    //Utilize os métodos toString de Cliente e Transação para montar a saída
    public String toString() {
        String saida = "clients:\n";
        for(Client client : repCli.values())
            saida += "- " + client + "\n";
        saida += "transactions:\n";
        for(Transaction tr : repTr.values())
            saida += "- " + tr + "\n";
        saida += "balance: " + balance + "\n";
        return saida.substring(0, saida.length() - 1); //remove last \n
    }
    int getBalance(){
        return balance;
    }
}

//!KEEP
class Solver{
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Agiota agiota = new Agiota(0);
        while(true){
            String line = scanner.nextLine();
            System.out.println("$" + line);
            String ui[] = line.split(" ");
            try{
                if(ui[0].equals("end")) {
                    break;
                } else if(ui[0].equals("init")) {
                    agiota = new Agiota(Integer.parseInt(ui[1]));
                } else if(ui[0].equals("addCli")) {
                    agiota.addClient(ui[1], Integer.parseInt(ui[2]));
                } else if(ui[0].equals("show")) {
                    System.out.println(agiota);
                } else if(ui[0].equals("kill")) {
                    agiota.kill(ui[1]);
                } else if(ui[0].equals("lend")) {
                    agiota.lends(ui[1], Integer.parseInt(ui[2]));
                } else if(ui[0].equals("receive")) {
                    agiota.receive(ui[1], Integer.parseInt(ui[2]));
                } else {
                    System.out.println("fail: comando invalido");
                }
            } catch (AgException ex) {
                System.out.println(ex.getMessage());
            }
        }
        scanner.close();
    }
}
//!OFF
